marconi-chain-index/src/Marconi/ChainIndex/Indexers/MintBurn.hs

```diff
@@ -293,56 +293,90 @@ data instance RI.StorableQuery MintBurnHandle
+    pure $ MintBurnResult $ do
+      TxMintEvent slotNo blockHeaderHash txAssets <- filteredMemoryEvents <> storedEvents
+      (txId, mintAssets) <- NE.toList txAssets
+      MintAsset policyId assetName quantity redeemerIx redeemerData <- NE.toList mintAssets
+      pure $ TxMintRow
+        slotNo
+        blockHeaderHash
+        txId
+        policyId
+        assetName
+        quantity
+        redeemerIx
+        redeemerData
 
+    where
+      filteredMemoryEvents :: [TxMintEvent]
+      filteredMemoryEvents = coerce $ fromRows $ filter rowFilter $ toRows =<< (coerce $ toList memoryEvents)
+
+      -- Applies every predicate to row, when all are True, then result is True.
+      rowFilter :: TxMintRow -> Bool
+      rowFilter row = let filters = mkRowPredicates query in all ($ row) filters
+
+      -- * Filter in-memory events
+
+      upToSlot :: Maybe C.SlotNo -> [TxMintRow -> Bool]
+      upToSlot maybeSlotNo =
+        maybe [] (\slotNo -> [\row -> _txMintRowSlotNo row <= slotNo]) maybeSlotNo
+
+      matchesAssetId :: C.PolicyId -> C.AssetName -> TxMintRow -> Bool
+      matchesAssetId policyId assetName row =
+        _txMintRowPolicyId row == policyId && _txMintRowAssetName row == assetName
+
+      isBurn :: TxMintRow -> Bool
+      isBurn row = _txMintRowQuantity row < 0
+
+      mkRowPredicates :: RI.StorableQuery MintBurnHandle -> [TxMintRow -> Bool]
+      mkRowPredicates = \case
+        QueryAllMintBurn maybeSlotNo -> upToSlot maybeSlotNo
+        QueryAllBurn     maybeSlotNo -> upToSlot maybeSlotNo <> [isBurn]
+        QueryByAssetId policyId assetName maybeSlotNo     -> upToSlot maybeSlotNo <> [matchesAssetId policyId assetName]
+        QueryBurnByAssetId policyId assetName maybeSlotNo -> upToSlot maybeSlotNo <> [matchesAssetId policyId assetName, isBurn]
+
+      -- * Filter sqlite events
+
+      mkSqliteConditions :: RI.StorableQuery MintBurnHandle -> ([SQL.Query], [NamedParam])
+      mkSqliteConditions = \case
+        QueryAllMintBurn slotNo ->
+          mkUpperBoundCondition slotNo
+        QueryByAssetId policyId assetName slotNo ->
+          mkUpperBoundCondition slotNo
+          <> mkAssetIdCondition policyId assetName
+        QueryAllBurn slotNo ->
+          mkUpperBoundCondition slotNo
+          <> (["quantity < 0"], [])
+        QueryBurnByAssetId policyId assetName slotNo ->
+          mkUpperBoundCondition slotNo
+          <> mkAssetIdCondition policyId assetName
+          <> (["quantity < 0"], [])
+
+      mkAssetIdCondition :: C.PolicyId -> C.AssetName -> ([SQL.Query], [NamedParam])
+      mkAssetIdCondition policyId assetName =
+        ( ["policyId = :policyId", "assetName = :assetName"]
+        , [":policyId" := policyId, ":assetName" := assetName]
+        )
+
+      mkUpperBoundCondition :: Maybe C.SlotNo -> ([SQL.Query], [NamedParam])
+      mkUpperBoundCondition maybeUpperBound = case maybeUpperBound of
+        Nothing -> ([], [])
```

Comment:
Since the function already returns a monomorphic list then this would indirectly still mean `[]`.

Reply:
oh no i meant instead of the whole `([],[])` stuff. And yes, `mempty` would mean the exact same thing but as we consider to append it in the end using `<>`, `mempty` in this specific case may be clearer (but I know that we may have different opinions on this so it's up to you to decide).

Reply:
If the type signature is monomorphic then being indirect makes it less clear, no? We had this discussion in the Utxo indexer where a direct implementation was replaced with one depending on Monad and Alternative instances. Also, how do you choose what to write if there were competing polymorphic ways to write the same thing, are they all better than the direct version?

Reply:
My reasoning is that one makes the structure explicit, while the other makes the purpose explicit. In one case, we insist on the fact that we return a pair of empty lists, in the other, we insist on the fact that it's the neutral element of this monoid. As the pair of lists isn't the nicer structure to read (to me), I prefer the latter.
But choosing between the two is definitely a matter of taste.
For example, I would probably prefer `""` over `mempty`, and definitely prefer `mempty` over `MkComplexDatatype [] "" 0 0 0` to give two extreme cases.

Reply:
Right, but since the return type is monomorphic then there is no "Monoid purpose", it's just that the type class definitions happen to match the current behavior. If the return type were `Monoid a => ... -> a`, then sure. With this (and the Utxo case as well) I'm arguing that using type class instances in the body of the function only adds indirection for no reason.